qutebrowser/browser/shared.py

```diff
@@ -326,3 +329,39 @@ def netrc_authentication(url, authenticator):
     authenticator.setPassword(password)
 
     return True
+
+
+def choose_file(multiple: bool):
+    """Select file(s) for uploading, using external command defined in config.
+
+    Args:
+        multiple: Should selecting multiple files be allowed.
+
+    Return:
+        A list of selected file paths, or empty list if no file is selected.
+        If multiple is False, the return value will have at most 1 item.
+    """
+    with tempfile.NamedTemporaryFile(delete=False) as tmpfile:
+        filename = tmpfile.name
+
+    if multiple:
+        command = config.val.fileselect.multiple_files.command
+    else:
+        command = config.val.fileselect.single_file.command
+
+    proc = guiprocess.GUIProcess(what='choose-file')
+    proc.start(command[0],
+               [arg.replace('{}', filename) for arg in command[1:]])
+
+    loop = qtutils.EventLoop()
+    proc.finished.connect(lambda _code, _status: loop.exit())
+    loop.exec_()
+
+    with open(filename, 'r', encoding='utf8') as tmpfile:
+        selected_file_list = tmpfile.read().splitlines()
+    os.remove(filename)
```

Comment:
I think it might be a bit safer to use the contextmanager to delete the file here by dropping `delete=False` (instead of deleting it manually). That way, we clean up if we, for example, throw an exception. I think this should still work as expected with the event loop stuff going on, but I'm not 100% sure.

Reply:
I don't know if that's possible. On Linux something like this works:

```
with tempfile.NamedTemporaryFile(mode='r') as tmpfile:
	name=tmpfile.name
	os.system('echo 12 > ' + name)
	print('content: ',tmpfile.read())
```

but I don't know if it's possible to have this process read from it and another process write to it at the same time on Windows/etc.

If we close and reopen it like this

```
with tempfile.NamedTemporaryFile(mode='r') as tmpfile:
	name=tmpfile.name
	print('name: ',name)
	tmpfile.close()
	os.system('echo 12 > ' + name)
	tmpfile=open(name,'r')
	print('content: ',tmpfile.read())
```

the file won't get deleted.

Reply:
oooh, I see, that's a bummer, I guess we can leave it as is...